### ISSUE TITLE
Notes iteration for WordPress 6.9

### ISSUE THREAD

**Issue by ellatrix** [WordPress/gutenberg#66377]
The initial PR: #60622 

## Next tasks

Summary: based on the new designs, we need two separate sidebars:

* One sidebar is coloured so it feels as an extension of the canvas. This "sidebar" should show whenever no other sidebar is open, such as inspector, global styles, second comment sidebar, etc. The comments should appear at the same height as the blocks. Only the top-level comment should be shown. Nested replies unfold when the top level comment is selected.
* A second sidebar that is a normal plugin sidebar: has toggle, white background. This sidebar should show _all_ comments (resolved or unresolved) in reverse chronological order. Nested replies still appear under the top level comment, and unfold when you select the top level comment.

## Next on design

We should also make progress on the design for inline comments. 

**First off, adding comments:**

![Image](https://github.com/user-attachments/assets/507dcb35-b177-4d8c-b902-00322c0be7f7)

Add on the block level, show an overlay marker, show those bubbles. As noted for other mockups, vertically aligning comments to text is meaningful, and in fact despite our conversation above about a "looking like the background" sidebar, there was an express desire to actually have those bubbles appear in context of the block. I recognize [this comment notes it to be hard to impossible](https://github.com/WordPress/gutenberg/issues/66377#issuecomment-2450010006), so this is a topic worth digging into, a little bit. 

In a way, I could think of these being positioned near the block same as how the block toolbar is positioned near the block, just hugging the right side of the screen. Let's brainstorm and see what's reasonable.

Here are some additional screens in the same category, allowing you to add multiple block level comments:

![Image](https://github.com/user-attachments/assets/394d4999-3a66-4b56-ba11-3070deef6eee)

Note in all of these, the _sidebar is toggled off_, because—after all—these comments are shown directly in the document, in context of the block.

**The sidebar, and its purpose**

That brings us to the sidebar, which can serve the purpose of showing _all comments added, in chronological order_:

![Image](https://github.com/user-attachments/assets/78e92e2a-3f09-4354-896f-647ca1241b88)

When the sidebar is opened, like in this mockup, all the comments in the canvas disappear. This sidebar is also where you might search comments, or see comments already resolved.

I'll leave this open for others to add items. Cc @annezazu @jasmussen @mtias

**Comment by hbhalodia** [comment]
Hi @ellatrix, I had created a quick PR for the issue 1, to use `TextareaControl` instead of `TextControl` component - https://github.com/WordPress/gutenberg/pull/66391

Thank You,

**Comment by annezazu** [comment]
@ellatrix  disregard my original comment. Edited to just have this as a sub-issue of the main overview issue and with a new title to make it clearer this is around what's next. Thanks for pulling this together!

**Comment by jasmussen** [comment]
Working on some mockups at the moment.

> In the main header toolbar button (that toggles the sidebar), could we maybe show a comment count so users know that comments have been made?

Not sure about this. It might eventually have value, but until it does, I'd tread carefully. Notably I would not want to add a notification number to the public sidebar API until we have a clearer vision for the future of this API. So I'd suggest this one being in the lower priority bucket.

**Comment by jasmussen** [comment]
A couple of mockups below, which look at next steps for inline commenting. Note, just like [last time](https://github.com/WordPress/gutenberg/pull/60622#issuecomment-2265158536), that your feedback and pushback, is valid. Despite their polished appearance, these are still mockups, subject to your feedback. Additionally, it is in the interest of such mockups to push the boundaries of what's possible, with the express intent to then pull back the pieces that are not reasonable. For example, contextual vertical positioning is likely going to be a headache, I've nevertheless included it, in case it's not.

**Adding comments, long term**

![Image](https://github.com/user-attachments/assets/d2706820-0952-456f-bddc-3d5c3daba665)

This mockup considers the feedback we've gotten, that the sidebar extension area (top right action button corner) is starting to be a little crowded. There's a separate task to improve this, through perhaps mimicking browser extension trays that can expand and collapse. I think I saw @richtabor mention this somewhere.

The solution for comments, here, is to create an addition tool for commenting. This is akin to Figma, and has previously been outlined in [a phase 3 post on workflows](https://make.wordpress.org/core/2023/07/04/workflows/). This would allow us a more deliberate context for comment-only tools, which might especially be useful for inline commenting.

In addition to the separate comment tool, it also adds a floating "Add comment" button, for contextual _inline-level_ comments.

**Adding comments, near term**

The separate tool and floating button is likely a ways out, dev-wise, with many smaller improvements worth prioritizing first, so as to stabililze this feature for 6.8. So a proposal for the near term is to keep using the sidebar API, _or something like it_, like so:

![Image](https://github.com/user-attachments/assets/fa43f716-2f51-4ba7-a02c-cdab1c0b9a72)

The flow outlined here is largely the same as what's shipping: comments sidebar, "Add comment" button in the ellipsis menu. Note a few details, though:

* The button to open the sidebar is present only if the document has one or more comments, otherwise it's hidden. That also removes the pinning/unpinning functionality. 
* Inline comments are possible. Select a piece of text, click the ellipsis/add-button comment, off you go.
* Note how adding comments is a modal action: adding a comment transfers focus there, traps tab, etc. The thicker blue outline indicates where focus is: first on the comment itself, then on the comment field.
* Note how the marker color changes depending on whether the block is selected, or not.

**Handling many comments, and commenters**

![Image](https://github.com/user-attachments/assets/45c43de4-064d-4516-81e9-98f78e819dd8)

Shown here, comment collapsing (a separator with "show 3 more comments").

**Adding multiple block-level comments to the same block**

![Image](https://github.com/user-attachments/assets/6c5b86a4-4fd6-4053-a02d-d684f1693e3c)

Shown here, adding multiple block-level comments to the same block (just click "Add comment" twice).

Let me know your thoughts.

[Figma](https://www.figma.com/design/CSlNU8T0EQ0dhLVIppNwFl/Annotations%2FComments%2FCollaborative-Editing?node-id=233-2752&t=n00OjPNgbKvzfkXH-1). Please see with "Dev mode" turned on, for additional comments.

**Comment by jasmussen** [comment]
Following up on the first draft mockups above, I have some additional ones, that are hopefully more immediately actionable, veering closer to what's shipping in the software. And ultimately it boils down to the following mockup of **simply adding block level comments**:

![Image](https://github.com/user-attachments/assets/e1ceee45-68ec-4fc9-88e8-f1c803ab8acd)

If we can get 90% towards this mockup, we'll be in a great place. CC: @ellatrix. Specifically we have some basic pieces that are important:

* Showing comments by default in the canvas, if present
* Aligning with blocks vertically
* Better signaling of blocks that have comments (the overlay/highlight/markers)

There are some additional details, noted contextually in the Figma file linked here at the end (turn on dev mode to see comments), but specifically it's about some modal-esque focus management when adding a comment, that seems important to make progress on.

In the same category, if we can get closer on the sidebar appearance to these mockups, that would help.

Additionally, it'd be nice to allow multiple comment threads on the same block:

![Image](https://github.com/user-attachments/assets/944d2b69-6b5e-4b4f-b36b-4be805891afb)

In case it's useful, here's a smaller map of different comment states:

![Image](https://github.com/user-attachments/assets/ff867707-4c0c-46ca-b400-53d296ae372f)


Finally, here's a sketch for resolving, and showing resolved comments:

![Image](https://github.com/user-attachments/assets/f93bb663-9ba4-4565-a879-5bd8675434bd)

I would think of this mockup as much less important/urgent than the important items mentioned above, just want to make sure that's covered visually. Kudos to @jarekmorawski for a collaboration on the mockups here.

[Figma with dev mode comments](https://www.figma.com/design/CSlNU8T0EQ0dhLVIppNwFl/Annotations%2FComments%2FCollaborative-Editing?node-id=233-2752&t=e58vICcSyyGZqCGn-1).

**Comment by ellatrix** [comment]
@jasmussen It's not possible to show block comment _in_ the canvas, since that's for content only. However, it should be possible to enlarge the canvas and position the comment UI over it (as popovers), much like the block toolbar. But this would require us to somehow "reserve" an area inside the canvas so no content can be below the comments. Perhaps a margin on the `html` element. Zoom-out mode has some similar margins. Although a margin on the `html` element would remove the canvas colour from the area. A margin on `body` could be tried to, but that starts to be more dangerous as it can clash with theme styles. It's tricky! 

An alternative idea would be to detect the colour of the canvas and fill the sidebar with it.

In both cases, comments still must be rendered in the side area and be repositioned to horizontally align with the blocks.

**Comment by jasmussen** [comment]
If you're referring to the color seeping in, I mean literally that, coloring the background of the sidebar to have the same color as global styles. Shown here, with a little bit of quick web-inspectoring:

![Image](https://github.com/user-attachments/assets/8c7dc402-8d07-41fa-b052-24939c1f0471)

**Comment by ellatrix** [comment]
@jasmussen Cool, that's absolutely (easily) possible. It will not work for classic themes, I guess I was thinking too much about making it perfect. :)

**Comment by jasmussen** [comment]
It's totally fine to fall back to a white background when a global definition is not present. Note also in the above _inspector mockup_ that the text and icon color would also need to be the text color of the theme, to have contrast.

**Comment by jarekmorawski** [comment]
While @jasmussen is busy polishing the near-term vision, I thought I'd explore the future if commenting was a separate mode in the editor.

The user would start by selecting the `Review` tool in the dropdown in the top left corner. I renamed it to Review because later, it would allow users to suggest changes and edits (more on that later).

![Image](https://github.com/user-attachments/assets/7d312375-9c31-4b3f-b6b6-cc8a40d6f8db)

We could display comments in a collapsed state next to the blocks they refer to. We’d highlight the associated text string or block when the user hovers over them.

![Image](https://github.com/user-attachments/assets/400d1c77-9446-48aa-be80-1d2c565d0084)

When people click the comment in the collapsed state, it expands and displays the full thread. The other comment 'bubbles' would be faded and pushed down to make room for the selected one. The reply input field would be immediately in focus when the commend expands. Clicking ‘Cancel’ would collapse the thread and all other bubbles would return to their places.

![Image](https://github.com/user-attachments/assets/576ac750-0af9-4013-b0b6-4d705aa4ea66)

At any moment the user can open the sidebar to view and interact with all comments and manage visibility settings (when the sidebar is open, the collapsed bubbles are not visible). In the sidebar, the user can decide to show/hide resolved comments as well as their own.
When `Show in the canvas` is unchecked, the comment bubbles from the previous screen are not shown in the editor.

![Image](https://github.com/user-attachments/assets/788fc745-6e43-44fc-84fa-ef69186285a9)

Clicking on any of the comments would highlight the text string or block in the canvas and show controls, including the reply field.

Now, looking even further into the future, **I explored what suggestions might look like**. We could display them similarly to comments, though we'd also need to provide all other editing tools. For instance, to replace text, people would just start typing and an automated message would show up in the Suggestions panel (it'd also be accounted for in the 'bubbles' shown in the canvas).

![Image](https://github.com/user-attachments/assets/1acfa066-0408-431c-a86e-09c1d8fa51a3)

The messaging would vary based on the action taken:

- Add [character/s or blocks]
- Remove [character/s blocks]
- Reordered [X] blocks
- Changed style of [block]

Users could approve and reject suggestions in the sidebar (hence why it's a separate mode) and filter by them in the sidebar.

![Image](https://github.com/user-attachments/assets/dc9c4451-f966-4e37-b537-0e0a10e55455)

Perhaps later on we'd introduce sharing permission and make it possible for devs to share links to the site editor where they'd only comment and update copy changes without interacting with blocks?

I'm excited about all the new opportunities multicollab will unlock. Let me know if these ideas help! cc @annezazu

**Comment by jasmussen** [comment]
Following on the lovely visioning by Jarek, and forgive me for the back and forth, here's yet another iteration, based on conversations and collaborations with @jarekmorawski as well (many kudos). Our goal remains to keep the iteration as small as we can for this next iteration, while still capturing some of the longer term goals.

**First off, adding comments:**

![Image](https://github.com/user-attachments/assets/507dcb35-b177-4d8c-b902-00322c0be7f7)

Still the same flow: add on the block level, show an overlay marker, show those bubbles. As noted for other mockups, vertically aligning comments to text is meaningful, and in fact despite our conversation above about a "looking like the background" sidebar, there was an express desire to actually have those bubbles appear in context of the block. I recognize [this comment notes it to be hard to impossible](https://github.com/WordPress/gutenberg/issues/66377#issuecomment-2450010006), so this is a topic worth digging into, a little bit. 

In a way, I could think of these being positioned near the block same as how the block toolbar is positioned near the block, just hugging the right side of the screen. Let's brainstorm and see what's reasonable.

Here are some additional screens in the same category, allowing you to add multiple block level comments:

![Image](https://github.com/user-attachments/assets/394d4999-3a66-4b56-ba11-3070deef6eee)

Note in all of these, the _sidebar is toggled off_, because—after all—these comments are shown directly in the document, in context of the block.

**The sidebar, and its purpose**

That brings us to the sidebar, which can serve the purpose of showing _all comments added, in chronological order_:

![Image](https://github.com/user-attachments/assets/78e92e2a-3f09-4354-896f-647ca1241b88)

When the sidebar is opened, like in this mockup, all the comments in the canvas disappear. This sidebar is also where you might search comments, or see comments already resolved.

Let me know anything you see here that's off, and I'm happy to course correct, and talk further.

**Comment by jeffpaul** [comment]
@jarekmorawski re:

> Clicking on any of the comments would highlight the text string or block in the canvas and show controls, including the reply field.

Perhaps should focus also go to the reply field for that comment as well?

**Comment by jarekmorawski** [comment]
@jeffpaul Absolutely. My proposal builds off of @jasmussen's design and doesn't impact his proposed interactions.

**Comment by annezazu** [comment]
I wanted to bring in some feedback from wordpress.org slack [from folks on the WSJ side of things](https://wordpress.slack.com/archives/C07NVJ51X6K/p1730724979643779?thread_ts=1730480863.941219&cid=C07NVJ51X6K) in reflection to what @jarekmorawski shared in particular:

>I especially like the Review idea, this is something we would warmly welcome. At the moment the block level commenting functionality is too high so we wouldn't be able to showcase this to the our Newsroom but when you get to text level especially if you can @ another user that would be something we could review.

Food for thought going forward from a real world newsroom 💥 I'll do a call for testing on Make Test in a few weeks when we are further along too.

**Comment by poojabhimani12** [comment]
> List the comments in the order in which the blocks appear.

@ellatrix - I'm having trouble retrieving the block client ID in site editor mode, which makes it challenging to arrange the comments in the correct order. Is there a way to locate the block client ID within the site editor?

**Comment by ppolo99** [comment]
> I wanted to bring in some feedback from wordpress.org slack [from folks on the WSJ side of things](https://wordpress.slack.com/archives/C07NVJ51X6K/p1730724979643779?thread_ts=1730480863.941219&cid=C07NVJ51X6K) in reflection to what [@jarekmorawski](https://github.com/jarekmorawski) shared in particular:
> 
> > I especially like the Review idea, this is something we would warmly welcome. At the moment the block level commenting functionality is too high so we wouldn't be able to showcase this to the our Newsroom but when you get to text level especially if you can @ another user that would be something we could review.
> 
> Food for thought going forward from a real world newsroom 💥 I'll do a call for testing on Make Test in a few weeks when we are further along too.

Love this. This was also highlighted by a couple of other large media customers - Having the option to review and even enable/disable the feature altogether was high up on their initial feedback.

**Comment by ndiego** [comment]
> At the moment the block level commenting functionality is too high so we wouldn't be able to showcase this to the our Newsroom but when you get to text level especially if you can @ another user that would be something we could review.

+1 to this sentiment. I immediately ran into this when doing some testing this evening. Block-level is a great start, though.

**Comment by ofadam** [comment]
Hello! A few observations/remarks from Gannett's perspective... we'd be thrilled to help demo/beta test any built-in commenting functionality as it moves ahead.

- I may be reading the discussion incorrectly, but our users would expect to be able to make further/continued edits to the post while having comments visible. The proposed "Review" mode would be nice to have, but I can't see users toggling it regularly just to interact with comments.

- We'd like the ability to see if a comment has been edited. That can help a user know that things may have changed since they last looked.

- We'd want the ability to add additional comments to a published post, since there may be continued revisions to a story after publication that need editor/reporter back-and-forth: https://github.com/WordPress/gutenberg/pull/66583

**Comment by jarekmorawski** [comment]
Thanks for the feedback, Adam. Much appreciated and very timely as I'm currently working on a new design that's exactly what you're describing. I'll share the details in a few days.

**Comment by bph** [comment]
I did a quick test of the current implementation of inline comments. Here are my notes: 

- When a comment is resolved there is no way to add new comments to a block or 'unresolve' it. Sometimes people have second thoughts to on their comments and want to amend but can't right now.
- Once a comment is resolved, it can't be open anymore.
- A second comment can only be a reply of the first one. So you can only have one thread of comments per block. It seems ok. Just was unclear.
- The comment count bubble on the **All Posts** screen is behaving oddly. The numbers of inline comments and on post comments are mixed and not accurate. 
- There is no way to access the comments unless you go into edit mode. It would be great to isolate a list of all comments on unpublished posts.
![Image](https://github.com/user-attachments/assets/5d938d80-dcaa-4a82-a43b-04a0079f1dd3)

**Comment by sethrubenstein** [comment]
We have just begun testing editor comments across our post types and with our editors. Will have some feedback in a few weeks.

**Comment by mtias** [comment]
@sethrubenstein we are picking this back up soon, curious if you got to write down some of that feedback :)

**Comment by sethrubenstein** [comment]
@mtias I do! We have been testing editor comments in a limited fashion, just amongst admins and a few key editors over the last few months and here are some initial impressions. Overall, everyone really liked the actual commenting experience. Leaving a comment is easy, and perhaps more importantly very performant in the editor. But there are some things that we think would make this ready for wider use, in order of what we think might be easiest to achieve down to hardest. 

- Indicating in the block list view when a block has a comment, ideally with a bubble showing how many.
  - In the block toolbar it would be nice to also have a bubble on the comment icon indicating how many open comments there are for that specific block
- Clicking on a comment in the sidebar should select the block in question in the editor
- Need ability to re-open resolved comments and surface resolved comments in the comments sidebar.
- Vertically aligning comments so that they appeared near the relevant block would also be welcome. 
- Ability to add attachments to editor comments.

Those are what we would see as critical to using editor comments more widely, especially in the page building/template experimenting process.

Then there's the longer much more complex functionality that would really let organizations like ours replace things like MS Word in our workflow for the editorial process... 

- Character level comments, the ability to highlight part of a string and add a comment to just that. Obviously, at least as I understand it, Block Bits will need to come much much further before that can even be attempted.

**Comment by mtias** [comment]
Awesome, thanks for the feedback, that's a great list! I think a couple of those things are part of the design spec, just not implemented yet (better indicators in toolbar, vertical alignment with blocks).

<img width="1600" alt="Image" src="https://github.com/user-attachments/assets/f54311f4-adbf-4459-99fe-b39b7d8bc2be" />

List view integration sounds like a great addition as well.

**Comment by fabiankaegy** [comment]
Adding a few thoughts here also as I've taken comments for a spin over the past week or so: 

 - Unresolving comments like others have already mentioned is something I think should be possible
 - Getting Email Notifications for comments would be such a nice addition. It's a very inefficient workflow to have to go to a post to see if someone has commented / responded to a comment. 
     - These notifications also should ideally happen in a "debounced" fashion. Not one email per comment immediately as they happen but instead with a slight delay but then grouped together 👀 _(But that probably is too much for an MVP here)_
 - @ mentions inside comments is another thing I immediately wanted to start doing. Would be cool to tie that into the email notifications like google docs does it. 
 - Inline comments feel sooo much more needed than block level comments. Maybe this is my personal bias. But when I am commenting on something it often is about a specific word / part of a sentence. Only being able to add a comment to the entire long paragraph is really limiting. It would be great to find a way to attach comments to inline text. We already have the mockups for it also: 
<img width="1922" height="1348" alt="Image" src="https://github.com/user-attachments/assets/79f22ca8-e75d-43e7-9f83-e82a01a5aeea" />

 - Tied to the inline comments, Actually being able to make suggestions which then can get accepted would be a huge win! 
 - +1 on adding an indicator for what block has comments in the list view & block toolbar
    - this could get cluttered though... So maybe it is time to add the "review" mode which then reveals all of this commenting functionality.

**Comment by joemcgill** [comment]
In preparation for the WordPress 6.9 roadmap, I'd like to take a pass at identifying the feature set from this discussion that needs to be finalized to move this feature out of experimental mode in order to try to include an initial version of this feature in the next major release of WP. This will also allow us to continue progressing the longer term vision and design exploration for this feature in #59445.

From what I can tell, much of the infrastructure and initial design language is already in place in the experimental version of this feature in the Gutenberg plugin as of today (v21.2.0), mostly based on the great work done in [the initial PR](https://github.com/WordPress/gutenberg/pull/60622).

To summarize this discussion and the feedback that has been given recently, here's my current understanding of what folks would like to see improved before releasing a non-experimental version of this feature:

**Needed improvements**

- Add the ability to reopen resolved comments
- Add comment indicators in the block toolbar and to the block list
- Highlight the related block when a comment is selected
- Add support for comments on published posts (See: #66583)

**Nice to have**

- Support for commenting on specific content within a block
- Improving the positioning of comments in relation to the related block
- Add support for rich text to comments (e.g., emoji, images, user mentions, etc)

**Long-term goals, not expected in this iteration**

- Introducing a new "Edit mode" tool for exposing commenting features
- Sending notifications for comments
- Add "apply suggestions" feature to comments

@ellatrix and @jasmussen does this generally match your expectations, or are there major requirements missing from this list that should be included? Once I've gotten feedback, we can create the necessary sub-tasks from this list for folks to work on. And to be explicit, I don't expect whatever initial issues we create to be 100% final, but want to help get a starting point to move this forward.

**Comment by mtias** [comment]
@joemcgill that list looks good. There were some pending tasks on the REST side for the comment type as well to follow up.

**Comment by ellatrix** [comment]
I think one major one missing is aligning the comments with blocks in the standard view

**Comment by joemcgill** [comment]
Thanks @ellatrix. That's what I had in mind with:

> Improving the positioning of comments in relation to the related block

**Comment by adamsilverstein** [comment]
> To summarize this discussion and the feedback that has been given recently, here's my current understanding of what folks would like to see improved before releasing a non-experimental version of this feature:

Thanks for this summary @joemcgill - this was a great starting point for me to catch up on the work and thought here so far.

One question that came up for me was _how do inline comments interact with revisions?_ My initial sense is that comments are anchored to a specific revision. Does that sound right?

Consider the following scenario:
* User A creates a new post and saves several drafts (which are stored as revisions), finally asking user B for review.
* User B opens the post leaves an inline comment on a block in the post, and interacts with User A in the comment thread.  
* User A restores a previous revision, before the comment was left.

What is the expected result?
My expectation is that the comment would "disappear", but would reappear if the revision with the comments was restored.

> There were some pending tasks on the REST side for the comment type as well to follow up.

@mtias where are these being tracked or worked on?

**Comment by annezazu** [comment]
>where are these being tracked or worked on?

Check this PR out: https://github.com/WordPress/gutenberg/pull/65181

Some more context on the [overarching tracking issue for block comments too](https://github.com/WordPress/gutenberg/issues/59445#issuecomment-2273413289).

**Comment by joemcgill** [comment]
> One question that came up for me was how do inline comments interact with revisions? 

Good question, @adamsilverstein. Thanks for describing the scenario you have in mind. I did some quick testing of how comments work with a similar scenario using Google docs and confirmed that comments there work as you described, and is how I would generally expect it to work as well. If the comment is made on a block and the post gets reverted to a previous version, then the comment would no longer be present. Technically, this is how the current implementation would have to work, since the parent block comment gets stored as an attribute of the block.

**Comment by joemcgill** [comment]
I've created sub-issues for all the improvements listed as "Must haves" to focus conversations and implementation efforts on those specific issues.

**Comment by t-hamano** [comment]
Please note that there are also a lot of accessibility issues that need to be fixed.  See: https://github.com/WordPress/gutenberg/issues/71249

**Comment by joemcgill** [comment]
For consistency, I've renamed this issue to "Block Commenting iteration for WordPress 6.9" to not confuse this broader effort with the desire to add the ability to attach comments to specific inline text within a block, listed as a nice to have [in this comment](https://github.com/WordPress/gutenberg/issues/66377#issuecomment-3120384455) as:

> Support for commenting on specific content within a block

**Comment by t-hamano** [comment]
Maybe these were already discussed in #60622, but I have a few concerns about the spec:

- A block with comments may be moved to another post or included as part of a pattern. As a result, `wp_comments.comment_post_ID` may not match the actual post.
- Deleting a block does not delete the linked comments from the database. Should comments be deleted along with a block? However, block deletions can be "undone" in the editor.
-  What happens to comments linked to blocks if the comment user is deleted?

**Comment by karthick-murugan** [comment]
@t-hamano - Here are my observations for the following 3 scenarios.

> * A block with comments may be moved to another post or included as part of a pattern. As a result, `wp_comments.comment_post_ID` may not match the actual post.

During testing, I observed that when a block is moved or copied to another post, the associated comments are not moved or copied to the target post. Similarly, when a block is converted into a pattern, the existing comments are not displayed in the sidebar (although they remain in the database). Only new comments can be added. Once the pattern is detached, the previously associated comments reappear in the sidebar.

> * Deleting a block does not delete the linked comments from the database. Should comments be deleted along with a block? However, block deletions can be "undone" in the editor.

When a block is deleted, the linked comments remain in the database. If the deletion is undone (e.g., using Ctrl+Z), the block reappears and the associated comments are once again visible in the sidebar. This suggests that comments are not permanently deleted with the block, which allows the undo functionality to work as expected.

> * What happens to comments linked to blocks if the comment user is deleted?

When a user who has left comments is deleted, the associated comments still appear in the sidebar since the comments is not deleted from the wp_comments table. Also the comment author’s name shows up since the wp_comments table has the comment_author field, which has the author name added.

**Comment by adamsilverstein** [comment]
One other behavior I noticed - if a block with a comment is split into two blocks (by hitting enter within a text block), the comment will become associated with both blocks (since both have the same comment id attribute). The result is clicking on either block will highlight the same comment. This makes it impossible to add comments to one of the two blocks separate from the other.

This may be the desired behavior  - I wanted to surface it here for wider discussion. Another option might be to link only the first of the two blocks to the comment after a split.

https://github.com/user-attachments/assets/8e3b29d8-f084-4e6e-9a0e-39fae83ad2ff

**Comment by karthick-murugan** [comment]
> One other behavior I noticed - if a block with a comment is split into two blocks (by hitting enter within a text block), the comment will become associated with both blocks (since both have the same comment id attribute). The result is clicking on either block will highlight the same comment. This makes it impossible to add comments to one of the two blocks separate from the other.
> 
> This may be the desired behavior - I wanted to surface it here for wider discussion. Another option might be to link only the first of the two blocks to the comment after a split.
> 
>  comment.block.split.mp4

I've implemented a fix for the block splitting issue in the Gutenberg editor. The problem was that when a block with an attached comment gets split into two blocks, both resulting blocks were inheriting the same blockCommentId attribute, causing the same comment to appear linked to both blocks. I've added a targeted fix in the __unstableSplitSelection function in packages/block-editor/src/store/actions.js that removes the blockCommentId from the tail block (second part) during splitting, while keeping it on the head block (first part). This ensures that only the first block retains the original comment, and the second block can accept new, independent comments. The fix is minimal - and doesn't break any existing functionality. I've tested it and confirmed that after splitting, each block now shows its own comment or no comment, allowing users to add separate comments to each split block as expected.

<img width="2430" height="1440" alt="Image" src="https://github.com/user-attachments/assets/4283dec9-a59e-47d4-9e9f-23ad97c30e6b" />

https://github.com/user-attachments/assets/1eb8c818-43df-4351-a57f-0ef4f8bdbad0

@adamsilverstein - Please let me know if the workflow is good to go, so that I can create a PR for it.

**Comment by adamsilverstein** [comment]
> @adamsilverstein - Please let me know if the workflow is good to go, so that I can create a PR for it.

The screencast looks good in terms of the alternative I described so yes, please go ahead and open a PR with what you have working. 

I'm still curious to see what others tracking this issue think the expected behavior should be...

**Comment by karthick-murugan** [comment]
> > [@adamsilverstein](https://github.com/adamsilverstein) - Please let me know if the workflow is good to go, so that I can create a PR for it.
> 
> The screencast looks good in terms of the alternative I described so yes, please go ahead and open a PR with what you have working.
> 
> I'm still curious to see what others tracking this issue think the expected behavior should be...

@adamsilverstein  Added a PR https://github.com/WordPress/gutenberg/pull/71511

**Comment by jeffpaul** [comment]
> I've implemented a fix for the block splitting issue in the Gutenberg editor. The problem was that when a block with an attached comment gets split into two blocks, both resulting blocks were inheriting the same blockCommentId attribute, causing the same comment to appear linked to both blocks. I've added a targeted fix in the __unstableSplitSelection function in packages/block-editor/src/store/actions.js that removes the blockCommentId from the tail block (second part) during splitting, while keeping it on the head block (first part). This ensures that only the first block retains the original comment, and the second block can accept new, independent comments. The fix is minimal - and doesn't break any existing functionality. I've tested it and confirmed that after splitting, each block now shows its own comment or no comment, allowing users to add separate comments to each split block as expected.

> I'm still curious to see what others tracking this issue think the expected behavior should be...

@karthick-murugan @adamsilverstein while I can see cases where this approach won't work well (e.g. the comment was intended for the tail block and is no longer attached to that block) it _may_ be better than having the same comment attached to both blocks.  I ran two tests in Google Docs and observed the following:

1. Commenting on a text fragment in-line and splitting that into different paragraphs maintains a singular comment across that text.  As we're not yet able to [comment inline](https://github.com/WordPress/gutenberg/issues/59445), this case is not yet relevant.
2. Commenting on an entire paragraph and splitting that into different paragraphs maintains a singular comment across those two paragraphs.

https://github.com/user-attachments/assets/f46c6439-64e6-4f40-b0d8-6ac9a9d25a34

My thoughts from this basic test are that: (1) when we're able to comment on a text fragment (e.g. inline commenting) that we'd want a singular comment to still reference that text fragment across multiple blocks, would not want to only pick the head block for the comment to stay attached to that part of the text fragment, and would not want to duplicate the comment for the text fragment on the now split out head and tail blocks; and (2) for the immediate case of commenting on an entire block that we'd want the comment to reference both the head and tail blocks, not just the head (or tail) block only, and not duplicating the comment for the head and tail blocks.

**Comment by adamsilverstein** [comment]
One challenge with linking the comment to both head and tail blocks is that it is not possible to add more than one comment thread to a block, however that's probably something we can tackle separately. I will create an issue to discuss further.

**Comment by t-hamano** [comment]
It turns out that the block commenting doesn't work in the template or template part, because they have string values for `postId`, such as `twentytwentyfive/home` or `twentytwentyfive/sidebar`. The `wp_comments` table requires the post ID as an integer.

I suggested disabling [the block commenting if the post ID is not an integer](https://github.com/WordPress/gutenberg/issues/71642), but I feel it's not a fundamental solution. 

I think the block commenting might be a part of post type support. Something like this:

```php
register_post_type(
	'page',
	array(
		// ...
		'supports' => array(
			'title',
			'editor',
			'author',
			'thumbnail',
			'page-attributes',
			'custom-fields',
			'comments',
			'revisions',
			'block-comment'
		),
		// ...
	)
);
```

Based on this support, we should be able to expose block comment functionality.:

```javascript
<PostTypeSupportCheck supportKeys="block-comments">
	<CollabSidebar />
</PostTypeSupportCheck>
```

**Comment by t-hamano** [comment]
There is about one month until Beta 1. Let's prioritize sub-issues with the Enhancement label, as basically enhancements will not be permitted after Beta 1.

**Comment by jeffpaul** [comment]
@t-hamano I went ahead and reviewed the labeling and any that didn't seem accurate I've commented on directly, otherwise I think we're set now on labels/types.

**Comment by t-hamano** [comment]
@Mamaduka @adamsilverstein, there have been so many PRs submitted regarding block comments that I'm a bit confused about the order in which to proceed.

My suggestion is below, what do you think?

1. First, merge the following three PRs:
  - [x] https://github.com/WordPress/gutenberg/pull/71862: This PR contains changes to the main comment components, so let's merge it first.
  - [x] https://github.com/WordPress/gutenberg/pull/71844: If E2e tests are merged first, regressions can be prevented in advance.
  - [x] https://github.com/WordPress/gutenberg/pull/71869/: Introducing new hooks can reduce a lot of code.
2. After that, I would like to propose removing the `focusThread` state. This state only visually indicates whether the thread is focused, but does not actually focus the thread. This state should strictly be `isOpen`, and the thread itself should be able to receive focus separately, regardless of the state. cc @joedolson 
3. Once we get this far, it will be easier to work on bug fixes and improvements, especially those related to accessibility. At the same time, we may need to add E2E tests to suit the purpose of each PR.

**Comment by Mamaduka** [comment]
@t-hamano, I think Enhancements or anything that can't be merged after beta should have priority. Having a good E2E test coverage would be helpful for us to iterate faster (#71844).

We should also re-evaluate REST API block comments. I think these will have slightly different permissions and conditions than normal comments.

> After that, I would like to propose removing the focusThread state. This state only visually indicates whether the thread is focused, but does not actually focus the thread. This state should strictly be isOpen, and the thread itself should be able to receive focus separately, regardless of the state.

Yes, this needs a little bit of re-thinking.

**Comment by t-hamano** [comment]
The PR that makes comment thread focusable and improve keyboard navigation: #71883

**Comment by adamsilverstein** [comment]
> We should also re-evaluate REST API block comments. I think these will have slightly different permissions and conditions than normal comments.

+1 Has this been discussed previously? What is the expected desired behavior for permissions? Do you think we need a new capability?

**Comment by Mamaduka** [comment]
@adamsilverstein, @t-hamano, create a separate issue where we can continue the discussion.

P.S. Currently checking what needs to be done and will post my findings there shortly.

**Comment by jeffpaul** [comment]
In reviewing the [list of remaining enhancements](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22Collaborative%20Workflows%22%20type%3AEnhancement) with @adamsilverstein @kadamwhite @poojabhimani12 @roseg43, we have the following updates/thoughts on the following:

- https://github.com/WordPress/gutenberg/issues/71393 was considered something to await post-beta 1 and potentially puntable, but in discussion with @annezazu is stressed as needing to be part of the baseline experience and thus continues to be targeted for WP 6.9
- https://github.com/WordPress/gutenberg/issues/71447 approach set, to be reviewed/tested
- https://github.com/WordPress/gutenberg/issues/71622 needs some review to confirm "Editorial Comments" as best naming method, confirmation on icon to use in post table list view, confirmation on whether they exist in the "Comments" left hand nav sidebar or in their own new "Editorial Comments" left hand nav sidebar
- https://github.com/WordPress/gutenberg/issues/71826 needed by Gutenberg 21.9 RC1 (aka Oct. 15th)
- https://github.com/WordPress/gutenberg/issues/71868 copy of "original block deleted" felt like best path forward to get this updated (iterate from there as design chimes in)
- https://github.com/WordPress/gutenberg/issues/71876 same questions as from https://github.com/WordPress/gutenberg/issues/71622 (see above)
- https://github.com/WordPress/gutenberg/issues/71891 design feedback needed
- https://github.com/WordPress/gutenberg/issues/71668 switched to `bug` but still targeted for resolution in WP 6.9
- https://github.com/WordPress/gutenberg/issues/71774 considered a `task` but still optimal to get into WP 6.9
- https://github.com/WordPress/gutenberg/issues/71906 switched to `task` but still optimal to get into WP 6.9

**Comment by jeffpaul** [comment]
FYI on a Discussion post that I've published on the topic of [Naming and Communication for Block Commenting in WP 6.9](https://github.com/WordPress/gutenberg/discussions/72014).

**Comment by jeffpaul** [comment]
In reviewing the list of remaining [enhancements](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Block%20Commenting%22%20type%3AEnhancement) and [bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Block%20Commenting%22%20type%3ABug) with @adamsilverstein @roseg43, we have the following updates/thoughts on the following:
* https://github.com/WordPress/gutenberg/discussions/72014#discussioncomment-14573475: current preferred naming of Annotate (verb) / Annotations (noun), but ideally need confirmation from Gutenberg leadership on naming decision to finalize labels/etc. across the feature 
* #71393: ready for review/merge, could use some core team attention on UX and finessing ahead of Gutenberg 21.9 RC / WP 6.9 Beta 1
* #71622: latest approach is to not show an extra "Editorial Comments" column or comment avatars on posts (as that's likely planned more as a presence indicator for real time collab), which means no attempt to alert folks to commenting in the admin in WP 6.9 and otherwise deferring to WP 7.0 where notifications may help resolve the concern here of "how do I/someone know there are comments needing review?"
* #71826: overall team to collaborate next week on tasks needed to handle this work ahead of Gb 21.9 RC
* #71447: character-limited filters recommended in https://github.com/WordPress/gutenberg/pull/71563#issuecomment-3369266700, work otherwise continuing (thanks @yashjawale!)
* #71891: awaiting reviews/testing
* @roseg43 to start digging into [remaining Accessibility bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Block%20Commenting%22%20type%3ABug%20label%3A%22%5BFocus%5D%20Accessibility%20(a11y)%22)

**Comment by Mamaduka** [comment]
The decisions for #71826 need to happen sooner rather than later. Syncing PHP changes into core takes time, and we don't want to affect the beta 1 schedule by late choices.

**Comment by jeffpaul** [comment]
@Mamaduka note that there's a PR up for review on #71826 should you have time to help review/guide the effort there: https://github.com/WordPress/gutenberg/pull/72181.

**Comment by jeffpaul** [comment]
Separately noting that @annezazu helped get us to a decision on naming (tldr "Notes"), see more here: https://github.com/WordPress/gutenberg/discussions/72014#discussioncomment-14629639.

**Comment by Mamaduka** [comment]
Thanks for the updates, @jeffpaul!

I like the name 🎶 😄.

I think that means we should also revise the user-facing string (internals will be updated as needed). Let's do it as part of #71964.

Do we also need to swap comment-related icons? cc @jasmussen

**Comment by Mamaduka** [comment]
Reading #72184, and I think we should also start referring to the feature as `notes` or `wp_notes` internally. This should reduce confusion.

We could start by changing the comment type name from `block_comment` to `note` or `wp_note`, because we won't be able to do it once the feature ships in core.

**Comment by jasmussen** [comment]
Asking here as it's not clear this is correct for a new issue, but what is the unread dot about here? 

<img width="175" height="75" alt="Image" src="https://github.com/user-attachments/assets/6cd47c78-af2f-403f-bba9-9bbbe86d7622" />

Mostly asking, because if a block has comments, they will both be visible in the canvas, and in the toolbar. If we need to delineate unread in addition to that, I'd put the unread dot in the bubble itself, and make it blue, like so:

<img width="1440" height="981" alt="Image" src="https://github.com/user-attachments/assets/6da423ea-f135-4371-9432-ce621f0eff4c" />

**Comment by jasmussen** [comment]
> Do we also need to swap comment-related icons?

My instinct is, let's keep the comment icon at least for the first outing of this:

* The comments themselves are literally bubbles in the margin of the canvas.
* Similar comment icons are used by other players in the space, for the same feature.

We can refine and make the icon more unique/branded as an enhancement, and can react to 6.9 feedback, definitely something to mull over, but I don't think it's a blocker.

**Comment by jeffpaul** [comment]
> We could start by changing the comment type name from block_comment to note or wp_note, because we won't be able to do it once the feature ships in core.

@Mamaduka would you prefer that in its on PR or bundled into another enhancement/task item for Notes (fka Block Commenting)?

**Comment by Mamaduka** [comment]
@jeffpaul, I'll create a PR to change the internals, most likely tomorrow morning.

**Comment by jeffpaul** [comment]
In reviewing the list of remaining [enhancements](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3AEnhancement) and [bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @adamsilverstein @roseg43 @poojabhimani12, we have the following updates/thoughts on the following:
* #71826 / #72181: prepared and awaiting merges on the below items
* #71622 / #71743: no longer desired functionality, removing from 6.9 project board + this 6.9 iteration issue (can add to 7.0 board later, but general problem will be covered by #72184 in the future)
* #72157 / #72214 + #72274: issue to be resolved by two separate PRs (tackling the two separate items from the issue), but this is otherwise considered a bug and will be re-classified as such
* #71447 / #71563: @yashjawale continuing to iterate based on feedback/testing, I'm reaching out to them to ensure they're able to resolve over next ~24hrs ahead of necessary merge for Gb 21.9 RC (otherwise may need to lean on @t-hamano to help finish up)
* #71393 / #72216: @adamsilverstein to resolve merge conflicts and coordinate to get merged in
* Renaming from "Block Commenting" to "Notes" to be done as part of #71964 or #71826 (@roseg43 @Mamaduka best for you both to coordinate on who's handling & as part of which PR)
* #71896: just noticed this is flagged as needed ahead of feature graduating out of experimental, tagging @adamsilverstein @poojabhimani12 @kadamwhite @roseg43 to see if someone can pick up this issue?

**Comment by t-hamano** [comment]
The following two enhancements have been merged:

- https://github.com/WordPress/gutenberg/pull/71563
- https://github.com/WordPress/gutenberg/pull/72216

I believe we have completed all of the essential enhancements for the 6.9 release, and we are ready to move on to terminology changes and stabilization.

**Comment by jeffpaul** [comment]
I believe we can now shift to finishing off #71826 to move the feature out from experimental within Gutenberg.  Are there any other enhancements, bugs, or tasks that folks see as necessary before moving the feature out of experiments?

**Comment by t-hamano** [comment]
Related: #72310

I missed the important point: following [the terminology decision](https://github.com/WordPress/gutenberg/discussions/72014#discussioncomment-14629639), the attribute name should also be changed from `attributes.metadata.commentId` to `attributes.metadata.noteId`, right?

**Comment by jeffpaul** [comment]
> I missed the important point: following [the terminology decision](https://github.com/WordPress/gutenberg/discussions/72014#discussioncomment-14629639), the attribute name should also be changed from `attributes.metadata.commentId` to `attributes.metadata.noteId`, right?

Probably best, yeah, but curious @adamsilverstein @Mamaduka @ellatrix how critical it is that non-editor/frontend UI also aligns perfectly to "notes"?

**Comment by t-hamano** [comment]
If we want to change attribute names, we'll need to do so before tomorrow's Gutenberg 21.9 RC release.

**Comment by joedolson** [comment]
I think that for long term code clarity and maintenance, it's fairly important.

**Comment by t-hamano** [comment]
> I missed the important point: following [the terminology decision](https://github.com/WordPress/gutenberg/discussions/72014#discussioncomment-14629639), the attribute name should also be changed from `attributes.metadata.commentId` to `attributes.metadata.noteId`, right?

To put this more concretely, it means what key the data is stored under.

```html
<!-- wp:paragraph { "metadata" : { "commentId" : 1 } } -->
<p>Hello World</p>
<!-- /wp:paragraph -->

<!-- Should we change it like this? -->

<!-- wp:paragraph { "metadata" : { "noteId" : 1 } } -->
<p>Hello World</p>
<!-- /wp:paragraph -->
```

This is purely an internal issue and is unlikely to be visible to users, but from a backwards compatibility perspective, it can be a bit of a hassle to change the key later.

**Comment by jeffpaul** [comment]
> This is purely an internal issue and is unlikely to be visible to users, but from a backwards compatibility perspective, it can be a bit of a hassle to change the key later.

Echoing @joedolson, I'd say yes to "Should we change it like this?" >> `noteId`.

**Comment by t-hamano** [comment]
I see, let's change the attribute name: https://github.com/WordPress/gutenberg/pull/72411

**Comment by jeffpaul** [comment]
In reviewing the list of remaining [enhancements](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3AEnhancement) and [bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @adamsilverstein, we have the following updates/thoughts on the following:
* 🎉 all enhancements have been merged, great work all! 🎉
* Adam to focus on [Core backport merge](https://github.com/WordPress/wordpress-develop/pull/10280#issuecomment-3421867843) today ahead of WP 6.1 beta 1 tomorrow
* #72440 /  #72454: focused on issues at bottom of screen; will open a follow-up issue for a larger solution where notes take up more space than sidebar allows (eg., top of screen, overflowing comments)
* #72444: Adam has draft PR, but working on testing and state handling before pushing up PR
* #72453: pinned sidebar has an overflow at the top of the sidebar and scrolling doesn’t align for new Notes, will be solved separately from this issue of overflow in floating sidebar
* #71668: will test post-Beta 1 for performance concerns to dig into
* #72195: waiting on details from @t-hamano on what’s additionally needed and whether others can help pitch in there
* #71927: not Notes-specific, but something related and being loosely tracked
* #71711 / #71728: JP will check-in with @peterwilsoncc on getting this updated to land during WP 6.9 beta / Gutenberg 22.0 (aka after other WP 6.9 enhancements are handled early this week)

**Comment by shimotmk** [comment]
I've created two issues regarding this note.

* #72520:
* #72521:

If possible, please add this to the WP6.9 tasks.

**Comment by MadtownLems** [comment]
I've been doing my best to follow all these threads as well as the beta-1 codebase, but can't figure out what was decided on how to disable block commenting / block notes. I saw talk of post-type-supports and of block supports. I would have thought block_editor_settings_all would have made sense, but don't see anything there.

What was decided for how one can disable block-commenting/notes?

**Comment by Mamaduka** [comment]
Hi, @MadtownLems 

The feature is controlled by the post types' `editor` supports a sub-property, and will be enabled for Posts and Pages. We'll publish the dev note as usual for the new features.

You can disable it by overriding the `editor` supports sub-properties.

```js
add_action( 'init', function() {
	add_post_type_support( 'post', 'editor' );
} );
```

**Comment by jeffpaul** [comment]
In reviewing the list of remaining [bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @adamsilverstein, @desrosj, @poojabhimani12, @roseg43 we have the following updates/thoughts on the following:

* #72440: review on whether to handle via current PR plus a follow-up or to try and resolve all related issues in a single PR (related to #72507); will need this fix regardless; tested lengthy comments + scrolling within Google Docs and it continues to scroll comments down past the end of the doc, but generally do not like approach for within Gutenberg; Adam to focus on an solution this week; @mamaduka ok to merge existing PR for Beta 2 as workable solution for now?
* #72548 & #72607 to be fixed in core via https://github.com/WordPress/wordpress-develop/pull/10405; ideally lands for Beta 2; #72607 // [Trac#64145](https://core.trac.wordpress.org/ticket/64145) "Notes should not appear in the context of comments" // [PR#10405](https://github.com/WordPress/wordpress-develop/pull/10405) "WP_Comment_Query: exclude the 'note' type, unless 'all' or 'note' type explicitly requested"
* Reviewed testing feedback from Jon
  * No keyboard shortcut for adding a Note (#72718)
  * Some confusion on where Notes icon appears in top toolbar, expected next to sidebar panel icon
  * Notes appearing as plugin in sidebar feels odd
  * first note on page clicking add Note from block toolbar doesn’t immediately place you into note area (#72720)
  * added note but didn’t save post, refreshed page and then showed note message that block was deleted (#72717)
  * odd snackbar warning (#72690 && [Trac#64153](https://core.trac.wordpress.org/ticket/64153))
  * add different UI handling in Notes sidebar for active notes, resolved notes, orphaned notes
  * Note icon in top toolbar almost too similar to Comments icon (#72730)
  * Noticed that Docs uses the icon to show the comment count
  * What happens if you delete a parent note in a thread, does it delete the entire discussion or just the top note?  Ideally only admins can delete others’ notes. (#72722)
  * any ability to restrict users from deleting comments or ability to undelete comments? (#72724)
  * add support for submitting the form using keyboard shortcuts, such as `⌘ + enter` (#72721)
* otherwise focusing on resolving bugs, open to folks helping to grab any that are not yet commented on / self-assigned

**Comment by Mamaduka** [comment]
> ok to merge existing PR for Beta 2 as workable solution for now?

We missed the Beta 2 window; packages are already synced. Let's focus on shipping the best possible experience for https://github.com/WordPress/gutenberg/issues/72440.

I think it would be helpful if we could expand on the remaining feedback items in their issues.

**Comment by t-hamano** [comment]
This issue has the `Needs Dev Note` label, but I'm wondering what kind of developer notes are needed. As I understand it, there's very little information that needs to be conveyed to developers.

If there's anything we should mention, it would be that this feature is part of the post type support?

```php
function custom_add_post_type_support() {
	$supports        = get_all_post_type_supports( 'my-post-type' );
	$editor_supports = array( 'notes' => true );

	// `add_post_type_support()` doesn't merge support sub-properties, so we explicitly merge it here.
	if ( is_array( $supports['editor'] ) && isset( $supports['editor'][0] ) && is_array( $supports['editor'][0] ) ) {
		$editor_supports = array_merge( $editor_supports, $supports['editor'][0] );
	}
	add_post_type_support( 'my-post-type', 'editor', $editor_supports );
}
add_action( 'init', 'custom_add_post_type_support' );
```

**Comment by MadtownLems** [comment]
> If there's anything we should mention, it would be that this feature is part of the post type support?

That's all I can think of, and I think including that bit in the dev note is an essential inclusion. Great snippet with comments. Thanks!

**Comment by jeffpaul** [comment]
@t-hamano if there's not a :alot: of dev documentation for us to highlight, then perhaps this gets rolled into the overarching Miscellaneous Editor dev note?

**Comment by jeffpaul** [comment]
Noting that @adamsilverstein will start a doc on things to document for Notes, depending on length of content we'll look to publish on its own (otherwise will roll into the Misc Editor note)

**Comment by jeffpaul** [comment]
In reviewing the list of remaining [bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @adamsilverstein @poojabhimani12 we have the following updates/thoughts on the following:

* [Trac#64165](https://core.trac.wordpress.org/ticket/64165): "At A Glance, You Can't Tell Which Blocks Have Notes And Which Don't"; JP to open this as an enhancement issue upstream here in GitHub
* https://github.com/WordPress/gutenberg/issues/72824#issuecomment-3479411326 (Joen’s note): I’ll get an enhancement issue opened for 7.0 on the “add note” on-select button and a separate one for note handling in Distraction Free Mode; probably best to hold off working on this for now while we focus more on resolving bugs in time remaining for 6.9
* https://github.com/WordPress/gutenberg/issues/72824#issuecomment-3480132598 (Aki’s note): as he confirmed, let’s see if we can get the "Disable the Add note button with some help text."—which could read: "Notes are disabled in distraction free mode." suggestion from Joen into a fix for 6.9
* https://github.com/WordPress/gutenberg/issues/72594#issuecomment-3437811169 (Joe’s note): I know Aki has flagged this as potential 6.9.1 candidate, but if we’re able to resolve this a11y issue for 6.9 that’ll make the overall experience a bit more polished for all users; Adam to ping Joe to confirm approach in the PR (separately from details noted in the issue)
* #72750: looks like some help may be needed on the linked PR #72872 with an odd test failure and resolving this will help avoid confusion on notes sidebar availability on posts  (can we remove strings after Beta1?); Adam to work on test failure, called out that it will remove strings (overall fine ahead of RC1)
* #72836: still need to resolve in the singular block “Edit as HTML” option, Karthick to work on a PR here
* #72892: Adam to work through feedback from JD on the linked PR
* #72862: existing UX is core WP behavior, one thing to test is when a parent Note is deleted that other Notes in that thread get marked as Trash and that core eventually garbage collects and deletes those (worth testing first with Comments and then identifying ideal handling for Comments and Notes which is likely at best a 7.0 item to check into)
* #72453: probably worth re-testing post-Beta3 to see if related updates in the pinned Notes sidebar resolves this, if not worth trying to resolve ahead of RC1
* #72717: may be worth documenting that you need to save a post before being able to add and save a note to a block, but for now may be a close as wontfix; tagging JD for input
* #72730: JP to ping Joen to try and get agreement to switch to his final icon option suggestion
<img width="734" height="140" alt="Image" src="https://github.com/user-attachments/assets/b158e064-757d-44ef-b0ab-3293258ef5d1" />

* https://github.com/WordPress/gutenberg/issues/72858#issuecomment-3474077280: worth resolving the last bits as highlighted by JD, may be resolved by #71969
* #72896: may already be resolved, effectively hiding or disabling “add note” for dynamic blocks
* #71668: some work has been done here, Adam to work on sub-issues for additional improvements on how to improve the UX overall
* Pooja to coordinate Multidots for additional testing once 6.9-beta3 is available, open issues as identified, craft PRs as time allows
* Figma link for latest designs: [Figma](https://www.figma.com/design/CSlNU8T0EQ0dhLVIppNwFl/Annotations-Comments-Collaborative-Editing?node-id=786-6328)
* need to confirm Noteworthy Contributors for Notes; as recommendation for About page, also for potential State of the Word callouts
* Adam to work on starting dev note doc for Notes

**Comment by t-hamano** [comment]
> if there's not a :alot: of dev documentation for us to highlight, then perhaps this gets rolled into the overarching Miscellaneous Editor dev note?

> Adam to work on starting dev note doc for Notes

Regarding the dev note: Probably the only thing that needs to be conveyed is that Notes is a type of post type support, and that this support can be added using `add_theme_support`.

However, there is also [a core ticket](https://core.trac.wordpress.org/ticket/64156) proposing improvements to the `add_post_type_support` function. Depending on the direction of this ticket, the code examples included in the developer notes may change slightly. cc @fabiankaegy

**Comment by desrosj** [comment]
Just in case it's not clear for everyone reading the latest update from @jeffpaul above (either now or in the future) the "JD" in summary is me.

**Comment by joedolson** [comment]
I assumed as much, but only because I knew it wasn't me... ;)

**Comment by t-hamano** [comment]
Since the RC1 release is coming soon, I would like to remind us again of the tasks that may need to be addressed in the 6.9 release. 

These are issues that have been added to the 6.9 project board and classified as bugs. Bugs that were punted to 6.9.1 are excluded. I will also quote @jeffpaul's review from [this comment](https://github.com/WordPress/gutenberg/issues/66377#issuecomment-3482317766).

- https://github.com/WordPress/gutenberg/issues/72717: may be worth documenting that you need to save a post before being able to add and save a note to a block, but for now may be a close as wontfix; tagging JD for input
- https://github.com/WordPress/gutenberg/issues/72862: existing UX is core WP behavior, one thing to test is when a parent Note is deleted that other Notes in that thread get marked as Trash and that core eventually garbage collects and deletes those (worth testing first with Comments and then identifying ideal handling for Comments and Notes which is likely at best a 7.0 item to check into)
- https://github.com/WordPress/gutenberg/issues/72858: worth resolving the last bits as highlighted by JD, may be resolved by https://github.com/WordPress/gutenberg/pull/71969

**Comment by t-hamano** [comment]
This issue was brought up in today's 6.9 bug scrub.

The release of RC1 is fast approaching, but let's try to address some of the tasks mentioned [here](https://github.com/WordPress/gutenberg/issues/66377#issuecomment-3509137768).

**Comment by jeffpaul** [comment]
In reviewing the [list of remaining bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @adamsilverstein @poojabhimani12 @desrosj @roseg43 we have the following updates/thoughts on the following:

* Aki highlighted remaining issues ideally resolved before RC1 (tomorrow) in https://github.com/WordPress/gutenberg/issues/66377#issuecomment-3509137768:
  * #72717: JP + @adamsilverstein concur on close as `wontfix`, @desrosj added that potentially adjusting autosave message to inform that it could also apply for Notes
  * #72862: @adamsilverstein confirmed that neither Comments or Notes delete children when a parent is deleted, could adjust for Notes but delicate on Comments as someone may already be depending on the current handling (or lack of handling) there; @desrosj’s preference to maintain orphaned notes in thread with placeholder “this note was deleted” for deleted note; @adamsilverstein opened https://github.com/WordPress/wordpress-develop/pull/10496 to handle
  * #72858: @karthick-murugan working to resolve e2e test failures on #73130 ahead of RC1
* #71668: remaining work on this punted to 7.0
* #71711: confirmed not specific to Notes and is Gutenberg-specific so will remove from iteration issue
* #72195: remaining items here punted to 7.0
* #72836: ready for review, @desrosj to test
* #72721: linked PR appears ready to merge, perhaps the polish here may be worth getting in for 6.9 and backporting before RC1?
* #73119: interesting use case, though I believe it was decided some time ago that the block would store the `noteID` and not vice versa so not sure how much we can assist here? @adamsilverstein to comment, likely just a bit of missing docs on this and to handle for this use case is a two-step flow to account for
* @adamsilverstein to commit https://github.com/WordPress/wordpress-develop/pull/10488 today
* #72184 opened to track notification-related enhancements for Notes as part of WP 7.0
* @desrosj concerned on circle colors around avatars potentially presenting as online presence indicators, opened #73144 to track that
* @adamsilverstein to work on dev note 
* @poojabhimani12 / Multidots to shift towards enhancements and 7.0 items

**Comment by jeffpaul** [comment]
FYI that I've created the #73260 issue and will move there any enhancement issues attached to this 6.9 issue that didn't get into the release (as well as any items in the "Punted to 7.0" queue once we ship WP 6.9 and the queue is finalized).

**Comment by jeffpaul** [comment]
In reviewing the [list of remaining bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20project%3AWordPress%2F249%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @adamsilverstein @poojabhimani12 @desrosj @roseg43 we have the following updates/thoughts on the following:

* #71668: will move to 7.0 and sub-issue back to 6.9 
* #71711: will remove from 6.9 tracking as no longer related to WP core (relates just to Gutenberg) and ping @Mamaduka  as reminder for review on related PR
* #72717: don’t technically need to save content when note added, meta could be saved; alternatively save content when a note is added (perhaps optimal handling)
* #72836: PR awaiting review, will ping @Mamaduka again (@adamsilverstein to also try to give a review)
* #72858: @t-hamano and @Mamaduka looking to merge, @desrosj to try and get to testing yesterday (_edit: this is now merged/done_)
* #73144: will leave open and await strong user feedback before determining an alternate path forward

**Comment by jeffpaul** [comment]
In reviewing the [list of remaining bugs](https://github.com/WordPress/gutenberg/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22%5BFeature%5D%20Notes%22%20type%3ABug) with @poojabhimani12 @desrosj @roseg43 we have the following updates/thoughts on the following

* #71668: JP to ping into Performance channel for assistance on testing here, potentially leveraging any tooling folks have available on that team; Pooja/Multidots tested scale into hundreds of notes, only getting first 100 notes back and may need to amend API to allow for alternate approach; had previously discussed requesting 100 first top level/parent notes and from there requesting additional child notes; perhaps iFramed editor may be able to help in some way here
* #72717: @westonruter’s latest comment a good insight on Notes in a read-only context (see #73418 as example of a read-only usage), but storing blockId or similar to a note has been rejected as an option during the early stages of the 6.9 cycle; uncertain why an auto save on the “add note” flow is not entertained; unlikely to be able to identify difference between “original block deleted” versus "unknown block”, best is perhaps a more generic phrase that covers both
* #72836: Adam confirmed Ophelia’s similar concern, otherwise PR appears to be best option for now, will ping @t-hamano and @Mamaduka for consideration of landing in 6.9
* #73144: will punt to 7.0 and await community feedback on whether /how we iterate
* #73412: probably more ideal, polished UX to maintain the expand/collapse state, but not such that its critical for 6.9 so punting to 7.0 to allow for community feedback on this and whether/how we iterate; Ophelia to check on picking this up this week
* #73473: perhaps throwing an AYS if they navigate away from the note form would help, otherwise maintaining the note message when they navigate back would be most ideal; Ophelia to check on picking this up this week
* Pooja/Multidots started in on 7.0 enhancements, some need design feedback, may look into scaling issue (#71668)
* Discussion on sharing feedback on how the cross-company collaboration on Notes when during WP 6.9 and making public appeals for others to contribute similarly on additional features looking for feature collaboration in 2026

**Comment by Mamaduka** [comment]
Should we close this in favor of the WP 7.0 tracking issue?

**Comment by t-hamano** [comment]
> - https://github.com/WordPress/gutenberg/issues/71711
> - https://github.com/WordPress/gutenberg/issues/73141

I have moved the above outstanding tasks to the 7.0 tracking issue.

> - https://github.com/WordPress/gutenberg/issues/71818

This final remaining task should be able to be closed in time for the 6.9 release.

**Comment by hanneslsm** [comment]
> [!TIP]
> Quick link to the 7.0 tracking issue: https://github.com/WordPress/gutenberg/issues/73260
